votor/src/consensus_pool/certificate_builder.rs

```diff
@@ -50,84 +100,124 @@ fn try_set_bitmap(bitmap: &mut BitVec<u8, Lsb0>, rank: u16) -> Result<(), Aggreg
     Ok(())
 }
 
-/// A builder for creating a `CertificateMessage` by efficiently aggregating BLS signatures.
-#[derive(Clone)]
-pub struct CertificateBuilder {
-    cert_type: CertificateType,
-    signature: SignatureProjective,
-    // For some certificates we need two bitmaps, for example, NotarizeFallback
-    // certificates have Notarize and NotarizeFallback votes, so we need two bitmaps
-    // to represent them. The order of the VoteType is defined in certificate_limits_and_vote_types.
-    // We normally put fallback votes in the second bitmap.
-    // The order of the VoteType is important, if you change it, you might interpret
-    // the bitmap incorrectly.
-    // Some certificates (like Finalize) only need one bitmap, then the second bitmap
-    // will be empty.
-    input_bitmap_1: BitVec<u8, Lsb0>,
-    input_bitmap_2: BitVec<u8, Lsb0>,
+/// Internal builder for creating [`Certificate`] by using BLS signature aggregation.
+#[allow(clippy::large_enum_variant)]
+enum BuilderType {
+    /// The produced [`Certificate`] will require only one type of [`VoteMessage`].
+    SingleVote {
+        signature: SignatureProjective,
+        bitmap: BitVec<u8, Lsb0>,
+    },
+    /// A [`Certificate`] of type NotarFallback or Skip will be produced.
+    ///
+    /// It can require two types of [`VoteMessage`]s.
+    DoubleVote {
+        signature: SignatureProjective,
+        bitmap0: BitVec<u8, Lsb0>,
+        bitmap1: Option<BitVec<u8, Lsb0>>,
+    },
 }
 
-impl CertificateBuilder {
-    pub fn new(cert_type: CertificateType) -> Self {
-        Self {
-            cert_type,
-            signature: SignatureProjective::identity(),
-            input_bitmap_1: BitVec::repeat(false, MAXIMUM_VALIDATORS),
-            input_bitmap_2: BitVec::repeat(false, MAXIMUM_VALIDATORS),
+impl BuilderType {
+    /// Creates a new instance of [`BuilderType`].
+    fn new(cert_type: &CertificateType) -> Self {
+        match cert_type {
+            CertificateType::NotarizeFallback(_, _) | CertificateType::Skip(_) => {
```

Comment:
I would still personally prefer `certificate_limits_and_vote_types` to be the single source of truth, but err okay...

We will add Genesis type soon and we need to add the knowledge it only contains one type of votes both in `certificate_limits_and_vote_types` and here. In this case it's still fine, only two places, but I had seen it being a slippery slope...

Reply:
the premise of this work and one of the core ideologies behind rust programming is to make such things explicit.

Adding genesis type will indeed make us change code in more than one place.  However, the compiler will help us in identifying all the places that are potentially impacted by this change so we can verify that the code properly works with it.

Imagine if somehow genesis did not fit into the existing model of one or two bitmaps with a single aggregate, and a newcomer like me only changed `certificate_limits_and_vote_types` and forget to make changes here.  Such coding styles help catch such problems.